runner/src/runner.rs

```diff
@@ -359,6 +364,59 @@ async fn run_binary(runner: &mut impl RunnerExt, stdout_mode: impl Into<Stdio>)
     Ok(child)
 }
 
+async fn run(runner: &mut impl RunnerExt) -> Result<(), Error> {
+    runner.subscribe_to_shutdown_signals()?;
+    // Create all directories for the `download_path` if they don't already exist.
+    fs::create_dir_all(&runner.download_path())?;
+    let release = runner.try_get_release(false).await?.expect("No current release");
+    // WARNING: This will overwrite any pre-existing binary with the same name
+    // TODO: Check if a release with the same version is already at the `download_path`
+    runner.download_binary(release).await?;
+
+    runner.run_binary().await?;
+
+    let mut loop_delay_counter = Duration::from_secs(0);
+    loop {
+        if runner.shutdown_flag().load(Ordering::Relaxed) {
+            // Terminate child process
+            runner.terminate_proc_and_wait()?;
+
+            // Break the loop to gracefully terminate the main process.
+            break;
+        }
+
+        tokio::time::sleep(OS_SIGNAL_PROCESSING_DELAY).await;
+        loop_delay_counter = loop_delay_counter.add(OS_SIGNAL_PROCESSING_DELAY);
+
+        // Only check for a new release if `BLOCK_TIME` has elapsed since the last check
+        if loop_delay_counter.lt(&BLOCK_TIME) {
+            continue;
+        }
+        // Reset the delay counter
+        loop_delay_counter = Duration::from_secs(0);
```

Comment:
I think there is a cleaner approach here but maybe @sander2 also has some thoughts. My suggestion would be to have `shutdown_flag()` return a channel and `select!` between that and the original `run()` implementation.

Reply:
I agree that I don't really like this loop based approach. I think using [signal-hook-tokio](https://docs.rs/signal-hook-tokio/latest/signal_hook_tokio/) would allow a cleaner implementation, where we can `select` on the `signals` stream and the `runner` future